### ISSUE TITLE
ESM based configuration file

### ISSUE THREAD

**Issue by thescientist13** [mochajs/mocha#5049]
**Is your feature request related to a problem or a nice-to-have?? Please describe.**
To continue the journey in [supporting ESM in Mocha](https://github.com/mochajs/mocha/releases/tag/v9.0.0), it would be nice to have support for ESM based configuration files.  Currently this is documented as a [known limitation](https://mochajs.org/#current-limitations)
> _[Configuration file](https://mochajs.org/#configuring-mocha-nodejs) can only be a CommonJS file (`.mocharc.js` or `.mocharc.cjs`)_

**Describe the solution you'd like**
Allow ESM based configuration files in projects, either with (`.mocharc.js`) or without  (`.mocharc.mjs`) a `type="module"` field in _package.json_
```js
export default {
  timeout: 30000
}
```

**Describe alternatives you've considered**
Only other current alternative in Node is CJS.

**Additional context**
N / A

**Comment by JoshuaKGoldberg** [comment]
I'm surprised there wasn't a previous issue filed for this already! 😄

**Comment by thescientist13** [comment]
Hey @JoshuaKGoldberg , saw you added the label for accepting PRs so was eager to dig into this. 

Was looking into mocha source and since Mocha config [lookup and parsing](https://github.com/mochajs/mocha/blob/master/lib/cli/config.js#L66) is synchronous, but ESM `import` is async, was wondering if you had any thoughts / guidance on what might be the best way to tackle this?

As far as I know, the only way to read an ESM file would be through `import`, e.g.
```js
const config = await import(pathToFileURL(filepath))
  .then(module => module.default));
```

Or is there another way to parse an ESM file to extract its imports I am not aware of?  🤔

**Comment by JoshuaKGoldberg** [comment]
I don't have many thoughts / guidance, sorry 😞. We the new maintainer team are still ramping up (#5027). I haven't looked into the implementation details for this or generally loading ESM.

Briefly:
* Other folks have dealt with the async<>sync issue by using `synckit`. That could work here... but I'd really prefer avoiding workarounds / tech debt if we don't truly need them.
* Changing `loadConfig` & co. would be a breaking change, but given that many other frameworks are having to do the same thing for ESM, I personally think that's reasonable.

**Comment by thescientist13** [comment]
> _Other folks have dealt with the async<>sync issue by using synckit. That could work here... but I'd really prefer avoiding workarounds / tech debt if we don't truly need them._

Yeah, I had seen some similar packages out there but like yourself, was still searching to see if there was a "dependency" free option out there.  In particular for synckit, involving in a Worker thread seems like a lot of overhead just read a config file.  🏋️ 

> _Changing loadConfig & co. would be a breaking change, but given that many other frameworks are having to do the same thing for ESM, I personally think that's reasonable._

I think this was my next avenue of exploration so good to know there is positive sentiment for that approach.  Let me play around with this and will report back if it looks promising.

----

I know this falls more on the side of adding more dependencies, but wasn't sure if you were aware of the tool [**CosmicConfig**](https://github.com/cosmiconfig/cosmiconfig), which aims to abstract away the whole config file lookup / parsing operations.  I haven't personally used it but a number of other big projects use it (like PostCSS I believe) so just figured I would put that out there.

**Comment by thescientist13** [comment]
Ok, got a [repo setup](https://github.com/thescientist13/mocha-esm-config) with [**patch-package**](https://www.npmjs.com/package/patch-package) (see the _patches/_ directory) and have a working implementation supporting both _.mjs_ and _.js_ + `type="module"` in _package.json_.  Just run `npm ci` and `npm test` to see it in action.

Aside from marking a couple functions as `async`, the main change here is in needing to [wrap _bin/mocha.js_ in an IIFE](https://github.com/thescientist13/mocha-esm-config/blob/master/patches/mocha%2B10.2.0.patch#L32) so that we can run a makeshift "top-level" `await` (since true top-level `await` is only available in modules)

Let me know your thoughts.

**Comment by JoshuaKGoldberg** [comment]
Awesome, thanks for the reference @thescientist13! This will be useful in triaging the feature request. ❤️ 

> aware of the tool [CosmicConfig](https://github.com/cosmiconfig/cosmiconfig)  

CosmicConfig is great! It'd be a bigger breaking change to use it in Mocha, but we'd certainly be open to at least an issue being filed to use it. I think we'd want to see a comparison of all the ways Mocha supports config files, all the ways CosmicConfig support config files, and any gaps between. We wouldn't want to lose support for an existing config form _if_ that config form is reasonable & used by users.

**Comment by thescientist13** [comment]
Thanks for taking a look @JoshuaKGoldberg . 🙏 

I'll start on this ESM PR then following my approach in the demo repo, and can follow up on the CosmicConfig issue after that.

**Comment by satyen95** [comment]
Hi @thescientist13 I'm also facing the same hassle of handling custom reporters & config files explicitly in `.cjs ` while my whole repo is maintained in `module JS` by default via package.json `type="module"`. Have you started to add support for this?

@JoshuaKGoldberg Is anyone contributing toward adding support for this?

**Comment by JoshuaKGoldberg** [comment]
Nobody's sent anything. You'd see it linked here if they did. 🙂

**Comment by nbkhope** [comment]
Is this issue getting traction? I'm currently migrating my projects to ESM and having a `.mocharc.js` file with `export default` triggers the error:

```
Error: Unable to read/parse [...]\.mocharc.js: Error [ERR_REQUIRE_ESM]: require() of ES Module [...]\.mocharc.js from [...]\node_modules\mocha\lib\cli\config.js not supported.
Instead change the require of .mocharc.js in [...]\node_modules\mocha\lib\cli\config.js to a dynamic import() which is available in all CommonJS modules.
```

**Comment by JoshuaKGoldberg** [comment]
Nobody's sent anything. You'd see it linked here if they did. 🙂

**Comment by thescientist13** [comment]
Sorry for going silent on this, but just wanted to call out to anyone eager to help out on this, that while I was able to achieve the 80 of the 80 / 20 rule, as expected, the last 20 is the trickiest part (classic 😅 ) and so thoughts / ideas on this last challenge can be found in my comment here based on my WIP at the time - https://github.com/mochajs/mocha/issues/5049#issuecomment-1935244476 .

----

That said @JoshuaKGoldberg do you think `require(ESM)` would just make this all "just work" out of the box now?  Maybe Mocha would have to add support for _.mjs_ extensions, but in theory, couldn all the heavy lifting be delegated to Node now?  🤔

**Comment by JoshuaKGoldberg** [comment]
I think so, yes! That would be great. The less the PR adding ESM support has to refactor existing code, the better.

**Comment by thescientist13** [comment]
Just a heads up I opened a PR for this based on `require(ESM)` - https://github.com/mochajs/mocha/pull/5397, but it does the caveat of only being supported in >= 20.19.0

Also realized another PR was opened as well for this but didn't notice that until after.

Happy to chat more in my PR about the Node stuff / tests / docs / etc if you like what you see 👍

**Comment by JoshuaKGoldberg** [comment]
> another PR was opened as well

Oh, yes 😬 we'd want to give priority to the person who sent it first. I've been on both sides of  _"second PR was merged first"_ and it can be a bad feeling either way. 

#5353's been in draft for a bit - I'll go over there and see if we can push it forward.
